demos/applications/apa102.cpp

```diff
@@ -0,0 +1,163 @@
+    led_strip.update(spi_frame.data);
+    hal::delay(clock, 500ms);
+    update_single<led_count>(blue, brightness, 2, spi_frame.data);
+    led_strip.update(spi_frame.data);
+    hal::delay(clock, 500ms);
+    update_single<led_count>(white, brightness, 3, spi_frame.data);
+    led_strip.update(spi_frame.data);
+
+    // update all LEDs at once
+    hal::delay(clock, 3s);
+    hal::print(console, "Updating all LEDS\n");
+    update_all<led_count>(rgb_array, brightness, spi_frame.data);
+    led_strip.update(spi_frame.data);
+
+    // cycle through RGB colors, start and end with red
+    hal::delay(clock, 3s);
+    hal::display::rgb_values rainbow = red;
+    hal::print(console, "Rainbow Cycle\n");
+    for (int i = 0; i <= 255; i++) {
+      // decrease red, increase blue
+      rainbow.red = 255 - i;
+      rainbow.blue = i;
+      std::array<hal::display::rgb_values, led_count> rainbow_array = {
+        { rainbow, rainbow, rainbow, rainbow }
+      };
+      update_all<led_count>(rainbow_array, brightness, spi_frame.data);
+      led_strip.update(spi_frame.data);
+      hal::delay(clock, 10ms);
+    }
+    for (int i = 0; i <= 255; i++) {
+      // decrease blue, increase green
+      rainbow.blue = 255 - i;
+      rainbow.green = i;
+      std::array<hal::display::rgb_values, led_count> rainbow_array = {
+        { rainbow, rainbow, rainbow, rainbow }
+      };
+      update_all<led_count>(rainbow_array, brightness, spi_frame.data);
+      led_strip.update(spi_frame.data);
+      hal::delay(clock, 10ms);
+    }
+    for (int i = 0; i <= 255; i++) {
+      // decrease green, increase red
+      rainbow.green = 255 - i;
+      rainbow.red = i;
+      std::array<hal::display::rgb_values, led_count> rainbow_array = {
+        { rainbow, rainbow, rainbow, rainbow }
+      };
+      update_all<led_count>(rainbow_array, brightness, spi_frame.data);
+      led_strip.update(spi_frame.data);
+      hal::delay(clock, 10ms);
+    }
+  }
+  hal::delay(clock, 50ms);
+}
+
+template<std::size_t PixelCount>
+void update_single(hal::display::rgb_values p_rgb,
+                   int p_brightness,
+                   size_t p_led_number,
+                   std::span<hal::byte> p_led_frames)
+{
+  size_t index = p_led_number * 4;
+  p_led_frames[index] = build_brightness_byte(p_brightness);
+  p_led_frames[index + 1] = p_rgb.blue;
+  p_led_frames[index + 2] = p_rgb.green;
+  p_led_frames[index + 3] = p_rgb.red;
+}
+
+template<std::size_t PixelCount>
+void update_all(std::span<hal::display::rgb_values> p_leds,
+                int p_brightness,
+                std::span<hal::byte> p_led_frames)
+{
+  for (size_t i = 0; i < p_leds.size(); i++) {
+    p_led_frames[i * 4] = build_brightness_byte(p_brightness);
+    p_led_frames[(i * 4) + 1] = p_leds[i].blue;
+    p_led_frames[(i * 4) + 2] = p_leds[i].green;
+    p_led_frames[(i * 4) + 3] = p_leds[i].red;
+  }
+}
+
+hal::byte build_brightness_byte(int p_brightness)
```

Comment:
```suggestion
hal::byte build_brightness_byte(unsigned p_brightness)
```

Otherwise you'll need to add an additional check for negative numbers in your function.